<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Journal</title>

<script src="site_libs/header-attrs-2.20/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/htmlwidgets-1.6.2/htmlwidgets.js"></script>
<link href="site_libs/leaflet-1.3.1/leaflet.css" rel="stylesheet" />
<script src="site_libs/leaflet-1.3.1/leaflet.js"></script>
<link href="site_libs/leafletfix-1.0.0/leafletfix.css" rel="stylesheet" />
<script src="site_libs/proj4-2.6.2/proj4.min.js"></script>
<script src="site_libs/Proj4Leaflet-1.0.1/proj4leaflet.js"></script>
<link href="site_libs/rstudio_leaflet-1.3.1/rstudio_leaflet.css" rel="stylesheet" />
<script src="site_libs/leaflet-binding-2.1.2/leaflet.js"></script>
<script src="site_libs/leaflet-providers-1.9.0/leaflet-providers_1.9.0.js"></script>
<script src="site_libs/leaflet-providers-plugin-2.1.2/leaflet-providers-plugin.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My lab Journal</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="Journal.html">Journal</a>
</li>
<li>
  <a href="Links.html">Links</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Journal</h1>

</div>


<p>This is my lab Journal, Here I’ve wrote some R scripts that I used
for the analysis of my data during my research, I hope You will find
this Usefull. The problems and tasks may not be in a chronological order
so feel free to surf the whole page in order to find what you’re looking
for !</p>
<p>Last compiled: 2023-05-05</p>
<div id="heatmap-for-antimicrobial-activity" class="section level1">
<h1>Heatmap for Antimicrobial activity</h1>
<div id="loading-the-necessary-packages" class="section level3">
<h3>Loading the necessary packages</h3>
<p>I will need the following packages to do most of the tasks</p>
<pre class="r"><code>library(reshape)
library(psych)
library(dplyr)
library(data.table)
library(ggplot2)
library(RSelenium)
library(ggplot2)</code></pre>
</div>
<div id="defining-our-data" class="section level3">
<h3>Defining our Data</h3>
<p>Here we could also have used the function read.csv or the “clipr”
packages to paste the table into R. The dataframe represents the results
of Antimicrobial activity testing of 39 bacterial isolates against five
indicator strains.</p>
<pre class="r"><code>tab &lt;- data.frame(&quot;ATCC 25922&quot; = c(1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0), &quot;ATCC 6269&quot; = c(1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0), &quot;E.coli 405&quot; = c(NA, NA, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, NA, NA, 1, 0, 0, 0, 0, 0, NA, NA, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, NA), &quot;E.coli BC36&quot; = c(NA, NA, NA, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, NA, NA, 1, 0, 0, 0, 0, 0, NA, NA, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, NA), &quot;MC4100&quot; = c(NA, 1, 0, 1, 1, 0, 0, NA, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, NA, 0, 0, 0, NA, NA, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, NA))
rownames(tab) &lt;- c(&quot;1288&quot;, &quot;32&quot;, &quot;108&quot;, &quot;15&quot;, &quot;443&quot;, &quot;645&quot;, &quot;478&quot;, &quot;758&quot;, &quot;1082&quot;, &quot;685&quot;, &quot;858&quot;, &quot;945&quot;, &quot;ET135&quot;, &quot;1068&quot;, &quot;ET87&quot;, &quot;940&quot;, &quot;ET172&quot;, &quot;199&quot;, &quot;966&quot;, &quot;680&quot;, &quot;942&quot;, &quot;854&quot;, &quot;756&quot;, &quot;123&quot;, &quot;207&quot;, &quot;ET124&quot;, &quot;457&quot;, &quot;357&quot;, &quot;ET176&quot;, &quot;445&quot;, &quot;105&quot;, &quot;43&quot;, &quot;629&quot;, &quot;257&quot;, &quot;1079&quot;, &quot;668&quot;, &quot;480&quot;, &quot;860&quot;, &quot;746&quot;)</code></pre>
</div>
<div id="reformat-table" class="section level3">
<h3>Reformat table</h3>
<p>Here we reformat the table in a way that could be used by ggplot
package</p>
<pre class="r"><code>tab_t &lt;- data.table::transpose(tab)
rownames(tab_t) &lt;- colnames(tab)
colnames(tab_t) &lt;- rownames(tab)
tab_t$indicator &lt;- rownames(tab_t)
tab_tm &lt;- melt(tab_t, id = &#39;indicator&#39;)
colnames(tab_tm) &lt;- c(&quot;indicator&quot;, &quot;strain&quot;, &quot;value&quot;)</code></pre>
</div>
<div id="heatmap-format-1" class="section level3">
<h3>Heatmap format 1</h3>
<p>Here we use the ggplot2 packages and geom_tile() to draw the heatmap.
coord_fixed are usefull to have square tiles.</p>
<pre class="r"><code>ggplot(tab_tm, aes(indicator, strain)) + geom_tile(aes(fill = factor(value)), colour = &quot;black&quot;) + theme(legend.title=element_text(colour = &quot;black&quot;, size = 12), axis.text.x = element_text(angle = 90, hjust=1, size=7)) + labs(title = &quot; Activity from supernatants&quot;, fill=&#39;Activity&#39;) + scale_fill_manual(values=c(&quot;white&quot;, &quot;black&quot;, &quot;grey&quot;), labels= c(&#39;Non active&#39;, &#39;Active&#39;, &#39;Not done&#39;)) +  coord_fixed()</code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
</div>
<div id="heatmap-format-2" class="section level3">
<h3>Heatmap format 2</h3>
<p>Here you will notice that the Y axis is reverted, in my case I prefer
to have the strain 1288 on top of the Y axis so I just add another layer
to invert it.</p>
<pre class="r"><code>ggplot(tab_tm, aes(indicator, strain)) + geom_tile(aes(fill = factor(value)), colour = &quot;black&quot;) + theme(legend.title=element_text(colour = &quot;black&quot;, size = 12), axis.text.x = element_text(angle = 90, hjust=1, size=7)) + labs(title = &quot;  &quot;, fill=&#39;Activity&#39;) + scale_fill_manual(values=c(&quot;white&quot;, &quot;black&quot;, &quot;grey&quot;), labels= c(&#39;Non active&#39;, &#39;Active&#39;, &#39;Not done&#39;)) +  coord_fixed() +  scale_y_discrete(limits=rev)</code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
</div>
</div>
<div id="heatmap-virulence" class="section level1">
<h1>Heatmap Virulence</h1>
<div id="loading-the-packages" class="section level3">
<h3>Loading the packages</h3>
<p>Load these packages first</p>
<pre class="r"><code>library(reshape)
library(psych)
library(dplyr)
library(data.table)
library(ggplot2)
library(RSelenium)
library(tidyverse)</code></pre>
</div>
<div id="import-and-reformat-the-data" class="section level3">
<h3>Import and reformat the Data</h3>
<p>Some Errors may Occur, check the class of your data while applying
functions, some functions work only on dataframes and some work only on
objects of class matrix.</p>
<pre class="r"><code>dfv &lt;- read.csv(file = &quot;C:/Users/User/OneDrive - Université Laval/OMICS/Virulence_Analysis/EC41_Virulence_factors__presence_absence.csv&quot;)
rownames(dfv) &lt;- dfv[,1] # assign the strains names to the rownames
dfv &lt;- subset (dfv, select = -X) # Remove the column &quot;-X&quot; containing the strains names
dfv_t &lt;- t(dfv) #transpose
rownames(dfv_t) &lt;- colnames(dfv)
colnames(dfv_t) &lt;- rownames(dfv)
dfv_t &lt;- as.data.frame(dfv_t) %&gt;% select(&quot;1288_S1&quot;, &quot;32_S2&quot;, &quot;108_S3&quot;, &quot;15_S4&quot;, &quot;443_S5&quot;, &quot;645_S6&quot;, &quot;478_S7&quot;, &quot;758_S8&quot;, &quot;1082_S9&quot;, &quot;685_S10&quot;, &quot;858_S11&quot;, &quot;945_S12&quot;, &quot;ET135_S13&quot;, &quot;1068_S14&quot;, &quot;ET87_S15&quot;, &quot;940_S16&quot;, &quot;ET172_S17&quot;, &quot;199_S18&quot;, &quot;966_S19&quot;, &quot;680_S20&quot;, &quot;942_S21&quot;, &quot;854_S22&quot;, &quot;756_S23&quot;, &quot;123_S24&quot;, &quot;207_S25&quot;, &quot;ET124_S26&quot;, &quot;457_S27&quot;, &quot;357_S28&quot;, &quot;ET176_S29&quot;, &quot;445_S30&quot;, &quot;105_S31&quot;, &quot;43_S32&quot;, &quot;629_S33&quot;, &quot;257_S34&quot;, &quot;1079_S35&quot;, &quot;668_S36&quot;, &quot;480_S37&quot;, &quot;860_S38&quot;, &quot;746_S39&quot;,  &quot;BC36_S40&quot;, &quot;Kp405_S41&quot;)   # reorder the data in that order
dfv_t$FV &lt;- row.names(dfv_t)
dfv_tm &lt;- melt(dfv_t)                          #Reshape package
colnames(dfv_tm) &lt;- c(&quot;FV&quot;,&quot;Strain&quot;,&quot;value&quot;)</code></pre>
</div>
<div id="draw-the-plot" class="section level3">
<h3>Draw the Plot</h3>
<pre class="r"><code>preferred.order &lt;- c(&quot;1288_S1&quot;, &quot;32_S2&quot;, &quot;108_S3&quot;, &quot;15_S4&quot;, &quot;443_S5&quot;, &quot;645_S6&quot;, &quot;478_S7&quot;, &quot;758_S8&quot;, &quot;1082_S9&quot;, &quot;685_S10&quot;, &quot;858_S11&quot;, &quot;945_S12&quot;, &quot;ET135_S13&quot;, &quot;1068_S14&quot;, &quot;ET87_S15&quot;, &quot;940_S16&quot;, &quot;ET172_S17&quot;, &quot;199_S18&quot;, &quot;966_S19&quot;, &quot;680_S20&quot;, &quot;942_S21&quot;, &quot;854_S22&quot;, &quot;756_S23&quot;, &quot;123_S24&quot;, &quot;207_S25&quot;, &quot;ET124_S26&quot;, &quot;457_S27&quot;, &quot;357_S28&quot;, &quot;ET176_S29&quot;, &quot;445_S30&quot;, &quot;105_S31&quot;, &quot;43_S32&quot;, &quot;629_S33&quot;, &quot;257_S34&quot;, &quot;1079_S35&quot;, &quot;668_S36&quot;, &quot;480_S37&quot;, &quot;860_S38&quot;, &quot;746_S39&quot;,  &quot;BC36_S40&quot;, &quot;Kp405_S41&quot;)
ggplot(dfv_tm, aes(x=FV, y= factor(Strain, levels = preferred.order))) + geom_tile(aes(fill = factor(value)), colour = &quot;white&quot;) + theme(legend.title=element_text(colour = &quot;black&quot;, size = 12), axis.text.x = element_text(angle = 90, hjust=1, size=6)) + labs(title=&quot;Virulome&quot;, size = 15) + labs(x=&quot;Virulence factor&quot;,y=&quot;Strain&quot;) + labs(fill=&#39;Genotype&#39;) + scale_fill_manual(values=c(&quot;white&quot;, &quot;black&quot;), labels= c(&#39;Absence&#39;, &#39;Presence&#39;)) +  scale_y_discrete(limits=rev) </code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
</div>
<div id="description-of-the-table" class="section level3">
<h3>Description of the table</h3>
<p>Get some stats such</p>
<pre class="r"><code>#describe(dfv)
descri &lt;- describe(dfv)
moyenne &lt;- subset (descri, select = &quot;mean&quot;)
moyenne$FV &lt;- row.names(moyenne)
view(moyenne)</code></pre>
</div>
<div id="reorder-virulence-factors-by-means" class="section level3">
<h3>Reorder Virulence Factors By Means</h3>
<pre class="r"><code>moyenne &lt;- moyenne[order(moyenne$mean), ]
good.order&lt;-moyenne$FV
ggplot(moyenne, aes(x = factor(FV, levels = good.order), mean)) + geom_col() </code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
</div>
</div>
<div id="resistance-phenotype" class="section level1">
<h1>Resistance Phenotype</h1>
<p>Our goal here is to visualize the resistance phenotype of the strains
of our collections by assigninging the colour <span
style="color:red">red</span> to resistant strains and <span
style="color:green">green</span> to sensitive ones.</p>
<p>We could also assign a colour for the intermediate resistance
profile, but it wont be the case here</p>
<div id="import-data" class="section level3">
<h3>Import Data</h3>
<pre class="r"><code>All_df &lt;- read.csv(file = &quot;C:/Users/User/OneDrive - Université Laval/OMICS/Paillasse/All_Data - Copy.csv&quot;) # This file contains most of the genomic data 
which(colnames(All_df) == &quot;AM&quot;) # 306 </code></pre>
<pre><code>## [1] 306</code></pre>
<pre class="r"><code>which(colnames(All_df) == &quot;Col&quot;) #322</code></pre>
<pre><code>## [1] 322</code></pre>
<pre class="r"><code>df&lt;- subset(All_df, select = c(306:322)) # I took only the columns related the antibiotic resistance profile
rownames(df) &lt;- All_df$gene # giving my rows relevant names
df_t &lt;- t(df) # transpose my table
rownames(df_t) &lt;- colnames(df) 
colnames(df_t) &lt;- rownames(df)

df_t &lt;- as.data.frame(df_t) %&gt;% select(&quot;1288_S1&quot;, &quot;32_S2&quot;, &quot;108_S3&quot;, &quot;15_S4&quot;, &quot;443_S5&quot;, &quot;645_S6&quot;, &quot;478_S7&quot;, &quot;758_S8&quot;, &quot;1082_S9&quot;, &quot;685_S10&quot;, &quot;858_S11&quot;, &quot;945_S12&quot;, &quot;ET135_S13&quot;, &quot;1068_S14&quot;, &quot;ET87_S15&quot;, &quot;940_S16&quot;, &quot;ET172_S17&quot;, &quot;199_S18&quot;, &quot;966_S19&quot;, &quot;680_S20&quot;, &quot;942_S21&quot;, &quot;854_S22&quot;, &quot;756_S23&quot;, &quot;123_S24&quot;, &quot;207_S25&quot;, &quot;ET124_S26&quot;, &quot;457_S27&quot;, &quot;357_S28&quot;, &quot;ET176_S29&quot;, &quot;445_S30&quot;, &quot;105_S31&quot;, &quot;43_S32&quot;, &quot;629_S33&quot;, &quot;257_S34&quot;, &quot;1079_S35&quot;, &quot;668_S36&quot;, &quot;480_S37&quot;, &quot;860_S38&quot;, &quot;746_S39&quot;) # reorder my columns in that order
df_t&lt;- df_t[colSums(!is.na(df_t)) &gt; 0]  # remove columns that have NA in it
df_t$ATB &lt;- row.names(df_t) # creating &quot;ATB&quot; column, necessary for the melt() to work
df_tm &lt;- melt(df_t)                        #Reshape package
colnames(df_tm) &lt;- c(&quot;ATB&quot;,&quot;Strain&quot;,&quot;value&quot;) # giving relevant names to the melted table
df_tm$ATB &lt;- factor(df_tm$ATB, levels = c(&quot;AM&quot;, &quot;AMC&quot;, &quot;FEP&quot;, &quot;CTX&quot;, &quot;FOX&quot;, &quot;CAZ&quot;, &quot;ATM&quot;, &quot;ETP&quot;, &quot;CIP&quot;, &quot;NA.&quot;, &quot;SXT&quot;, &quot;C&quot;, &quot;TE&quot;, &quot;GM&quot;, &quot;K&quot;, &quot;S&quot;, &quot;Col&quot;))
ggplot(df_tm, aes(ATB, Strain)) + geom_tile(aes(fill = factor(value), colour = c(&quot;red&quot;,&quot;green&quot;,white) ), colour = &quot;white&quot;) + theme(legend.title=element_text(colour = &quot;black&quot;, size = 12), axis.text.x = element_text(angle = 45, hjust=1, size=8)) + labs(title=&quot;Antibiotic resistance Profile&quot;, size = 15) + labs(x=&quot;Antibiotic&quot;,y=&quot;strain&quot;) + labs(fill=&#39;Phenotype&#39;) + scale_fill_manual(values=c(&quot;green&quot;,&quot;red&quot;, &quot;red&quot;), labels= c(&#39;Sensitive&#39;, &#39;Resistant&#39;, &#39;Resistant&#39;))</code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
</div>
</div>
<div id="resistance-heatmap" class="section level1">
<h1>Resistance Heatmap</h1>
<div id="loading-the-necessary-packages-1" class="section level3">
<h3>Loading the necessary packages</h3>
<pre class="r"><code>library(reshape)
library(psych)
library(dplyr)
library(data.table)
library(ggplot2)
library(RSelenium)
library(ggplot2)</code></pre>
</div>
<div id="importing-and-transforming-data" class="section level3">
<h3>Importing and transforming data</h3>
<pre class="r"><code>R_Genes &lt;- read.csv(file = &quot;C:/Users/User/OneDrive - Université Laval/OMICS/Résistance_Analyse/R_Genes.csv&quot; )

rownames(R_Genes) &lt;- R_Genes$gene

R_Genes &lt;- subset(R_Genes, select = -c(1))

R_Genes &lt;- subset(R_Genes, select=-c(which(colnames(R_Genes)==&quot;BC36_S40&quot;))) # Remove this sample

R_Genes &lt;- subset(R_Genes, select=-c(which(colnames(R_Genes)==&quot;Kp405_S41&quot;))) # Remove this sample


R_Genes_t &lt;- t(R_Genes)

R_Genes_tm &lt;- reshape2::melt(R_Genes_t)

colnames(R_Genes_tm) &lt;- c(&quot;strain&quot;,&quot;gene&quot;,&quot;value&quot;)

p&lt;-ggplot(R_Genes_tm, aes(strain, gene)) + geom_tile(aes(fill = factor(value)), colour = &quot;black&quot;) + theme(legend.title=element_text(colour = &quot;black&quot;, size = 3), axis.text.x = element_text(angle = 90, hjust=1, size=9), axis.text.y = element_text(size = 3.5)) + labs(title=&quot;Resistomes&quot;, size = 100) + labs(x=&quot;Strains&quot;,y=&quot;Resistance genes and mutations&quot;) + labs(fill=&#39; &#39;, ) + scale_fill_manual(values=c(&quot;black&quot;, &quot;red&quot;, &quot;green&quot;), labels= c(&#39;Absence&#39;, &#39;strict hit&#39;, &#39;perfect hit&#39;))

p #view plot</code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
</div>
<div id="saving-the-plot" class="section level3">
<h3>Saving the plot</h3>
<pre class="r"><code>ggsave(p, path = &quot;C:/Users/User/OneDrive - Université Laval/Article Drafts/1st article Tables and Figures/&quot;, filename = &quot;heatmap.pdf&quot;)</code></pre>
</div>
<div id="changing-the-order-of-the-strains" class="section level3">
<h3>Changing the order of the strains</h3>
<p>We want to put a custom order before redrawing the graph</p>
<pre class="r"><code>good_order &lt;- c(&quot;X1288_S1&quot;,&quot;X32_S2&quot;,&quot;X108_S3&quot;,&quot;X15_S4&quot;,&quot;X443_S5&quot;,&quot;X645_S6&quot;,&quot;X478_S7&quot;,&quot;X758_S8&quot;,&quot;X1082_S9&quot;,&quot;X685_S10&quot;,&quot;X858_S11&quot;,&quot;X945_S12&quot;,&quot;ET135_S13&quot;,&quot;X1068_S14&quot;,&quot;ET87_S15&quot;,&quot;X940_S16&quot;,&quot;ET172_S17&quot;,&quot;X199_S18&quot;,&quot;X966_S19&quot;,&quot;X680_S20&quot;,&quot;X942_S21&quot;,&quot;X854_S22&quot;,&quot;X756_S23&quot;,&quot;X123_S24&quot;,&quot;X207_S25&quot;,&quot;ET124_S26&quot;,&quot;X457_S27&quot;,&quot;X357_S28&quot;,&quot;ET176_S29&quot;,&quot;X445_S30&quot;,&quot;X105_S31&quot;,&quot;X43_S32&quot;,&quot;X629_S33&quot;,&quot;X257_S34&quot;,&quot;X1079_S35&quot;,&quot;X668_S36&quot;,&quot;X480_S37&quot;,&quot;X860_S38&quot;,&quot;X746_S39&quot;)

p1&lt;-ggplot(R_Genes_tm, aes(x=factor(strain, levels = good_order ), gene)) + geom_tile(aes(fill = factor(value)), colour = &quot;black&quot;) + theme(legend.title=element_text(colour = &quot;black&quot;, size = 3), axis.text.x = element_text(angle = 90, hjust=1, size=9), axis.text.y = element_text(size = 3.5)) + labs(title=&quot;Resistomes&quot;, size = 100) + labs(x=&quot;Strains&quot;,y=&quot;Resistance genes and mutations&quot;) + labs(fill=&#39; &#39;, ) + scale_fill_manual(values=c(&quot;black&quot;, &quot;red&quot;, &quot;green&quot;), labels= c(&#39;Absence&#39;, &#39;strict hit&#39;, &#39;perfect hit&#39;))

p1</code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
</div>
</div>
<div id="get-the-genotype-from-this-data" class="section level1">
<h1>Get the genotype from this data</h1>
<p>In the previous code we have been able to draw a heatmap of the
strains. In the case that i want to have the Profile of each strain I
have to proceed as following.</p>
<div
id="load-the-data-and-subset-the-antibiotic-resistance-profile-section"
class="section level3">
<h3>Load the data and subset the Antibiotic resistance profile
section</h3>
<pre class="r"><code>All_df &lt;- read.csv(file = &quot;C:/Users/User/OneDrive - Université Laval/OMICS/Paillasse/All_Data - Copy.csv&quot;) # This file contains most of the genomic data 
which(colnames(All_df) == &quot;AM&quot;) # 306 </code></pre>
<pre><code>## [1] 306</code></pre>
<pre class="r"><code>which(colnames(All_df) == &quot;Col&quot;) #322</code></pre>
<pre><code>## [1] 322</code></pre>
<pre class="r"><code>antibiogram&lt;- subset(All_df, select = c(306:322)) # I took only the columns related the antibiotic resistance profile
rownames(antibiogram) &lt;- All_df$gene # giving my rows relevant names
antibiogram&lt;- t(antibiogram)
antibiogram &lt;- antibiogram[, !colSums(is.na(antibiogram)), drop = FALSE]# remove columns that have NA in it
antibiogram&lt;- t(antibiogram)</code></pre>
</div>
<div id="declaring-an-empty-dataframe-to-store-results"
class="section level3">
<h3>Declaring an empty Dataframe to store results</h3>
<pre class="r"><code>return_final&lt;-data.frame()</code></pre>
</div>
<div id="design-a-for-loop" class="section level3">
<h3>Design a for loop</h3>
<p>For loop to read into matrix and then fill a new table with the
resistance profile</p>
<p>The following lines will :</p>
<ol style="list-style-type: decimal">
<li><p>Read the table line by line with a first for loop. each line
(strain) will be assigned to a temporary table “tmp”.</p></li>
<li><p>Subset only the columns (antibiotics) that have a value equal to
1 (resistant phenotype) and assign it to another temporary table
“tmp1”.</p></li>
<li><p>Declare an empty string vector here called “profile”.</p></li>
<li><p>Fill the empty vector with the column names of the table “tmp1”
that contains only column names of antibiotics to which a strain is
resistant.</p></li>
<li><p>Fill a temporary table with the results.</p></li>
<li><p>Bind all the temporary tables into a final results table</p></li>
</ol>
<pre class="r"><code>data &lt;- as.data.frame(antibiogram)
for (i in row.names(data)) {        
  tmp&lt;-data[c(i),]
  tmp1 &lt;- subset(tmp, select = c(tmp==1))
  nbr&lt;- print(length(tmp1))
  profile&lt;-as.character(&quot; &quot;)
  for (j in colnames(tmp1)){
    profile&lt;- paste0(profile,&quot;/&quot;, print(j))
  }
  return_tmp &lt;- data.frame(&quot;Strain&quot; = i, &quot;profile&quot; = profile, &quot;Resistant to&quot; = nbr, stringsAsFactors = F)
  return_final&lt;-rbind(return_final, return_tmp)
}</code></pre>
</div>
<div id="view-the-results" class="section level3">
<h3>View the Results</h3>
<pre class="r"><code>view(return_final)</code></pre>
</div>
<div id="save-the-results-in-a-csv-table" class="section level3">
<h3>Save the results in a csv table</h3>
<p>In order to keep these results</p>
<pre class="r"><code>write.csv(return_final, file = &quot;/Users/User/OneDrive - Université Laval/OMICS/Paillasse/Antibiogramme_profile.csv&quot; )</code></pre>
</div>
</div>
<div id="description-for-a-table" class="section level1">
<h1>Description for a table</h1>
<pre class="r"><code>table &lt;- describe(df)
view(table)
ggplot(table, aes(x=vars, y=mean)) + geom_bar(stat = &#39;identity&#39;)</code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-21-1.png" width="672" /></p>
<p>#Mapping samples origins</p>
<p>I want to show where my samples come from on a map</p>
<div id="load-necessary-packages" class="section level3">
<h3>Load necessary packages</h3>
<pre class="r"><code>library(leaflet)
library(tmaptools)</code></pre>
</div>
<div id="preparing-data" class="section level3">
<h3>Preparing Data</h3>
<p>Im going to try to take GPS localisation of the samples. Try to give
them different names according to the names in this server <a
href="https://nominatim.openstreetmap.org/"
class="uri">https://nominatim.openstreetmap.org/</a></p>
</div>
<div id="rendering-the-map" class="section level3">
<h3>Rendering the map</h3>
<pre class="r"><code>simple&lt;- data.frame(&quot;Samples&quot;=preferred.order, ID= c(&quot;Nabeul&quot;, &quot;ksour essef&quot;, &quot;mellassine&quot;, &quot;bou merdes&quot;, &quot;sidi bou ali&quot;, &quot;kemicha&quot;, &quot;grombalia&quot;, &quot;Gouvernorat Siliana&quot;, &quot;Jandouba&quot;, &quot;Gouvernorat Monastir&quot;, &quot;Chela&quot;, &quot;Sijoumi&quot;, &quot;Sfax&quot;, &quot;Mahres&quot;, &quot;Stade Municipal Menzel Chaker&quot;, &quot;El Habibia&quot;, &quot;Manouba&quot;, &quot;La Marsa&quot;, &quot;تونس العاصمة&quot; , &quot;Municipalité De Beb Lakwes&quot;, &quot;Bizerte&quot;, &quot;Chela&quot;, &quot;El Aouina&quot;, &quot;mellassine&quot;, &quot;Gouvernorat Béja&quot;, &quot;Zaghouan&quot;, &quot;Bir Ali Ben Khalifa&quot;, &quot;Chela&quot;,&quot;Gouvernorat Mahdia&quot;, &quot;Hayy Ibn Khaldun&quot;, &quot;mellassine&quot;, &quot;Délégation Ksour Essef&quot;, &quot;تونس العاصمة&quot;, &quot;Rue Ibn Charaf&quot;, &quot;Intilaka&quot;, &quot;Rue Touhami Amar&quot;, &quot;Agareb&quot;, &quot;Aïn Draham&quot;, &quot;Rue Beb Jdid&quot;, NA, NA))
cities&lt;- simple$ID[1:39] # cut the names of the cities and remove the last values which are NAs
query &lt;- geocode_OSM(cities)
m &lt;- leaflet() %&gt;%
     addTiles(urlTemplate = &quot;https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png&quot;) %&gt;%    addMarkers(query$lon, query$lat, label = query$query,  icon = makeIcon(iconUrl =&quot;https://fanblairfarm.co.uk/wp-content/uploads/2020/11/chicken-icon-2.png&quot;, iconWidth = 17, iconHeight= 17)) </code></pre>
</div>
<div id="try-this-format" class="section level3">
<h3>Try this format</h3>
<pre class="r"><code>m %&gt;% addProviderTiles(providers$MtbMap) %&gt;%
  addProviderTiles(providers$Stamen.TonerLines,
    options = providerTileOptions(opacity = 0.35)) %&gt;%
  addProviderTiles(providers$Stamen.TonerLabels)</code></pre>
<div class="leaflet html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-45aabe3362cb400e2a9b" style="width:672px;height:480px;"></div>
<script type="application/json" data-for="htmlwidget-45aabe3362cb400e2a9b">{"x":{"options":{"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}}},"calls":[{"method":"addTiles","args":["https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",null,null,{"minZoom":0,"maxZoom":18,"tileSize":256,"subdomains":"abc","errorTileUrl":"","tms":false,"noWrap":false,"zoomOffset":0,"zoomReverse":false,"opacity":1,"zIndex":1,"detectRetina":false}]},{"method":"addMarkers","args":[[36.4512897,35.4162835,36.79515,35.461005,35.9606191,36.5044991,36.5984231,35.9715323,36.494946,35.60547975,36.705375,36.7769937,34.7232387,34.5218412,34.9693859,36.8108651,36.7624363,36.8790882,36.8002068,36.80647135,37.2720905,36.705375,36.852609,36.79515,36.74623035,36.3319504,34.7267806,36.705375,35.3409383,36.82794,36.79515,35.35660215,36.8002068,36.7989826,36.8391913,35.8156262,34.7267589,36.77393155,36.8597626],[10.7355915,10.9987172,10.1515887182475,10.7091095,10.4821146467179,10.6766828,10.4990402,9.3577128756367,8.776369,10.7876947685139,10.2982458529412,10.1091852321074,10.3358474256388,10.4984421,10.3738282683642,9.95370955407237,9.83361907826277,10.327678,10.1857757,10.1641321520558,9.8708565,10.2982458529412,10.2659363,10.1515887182475,9.20323283692005,10.0452999718776,10.1014808309275,10.2982458529412,10.6041957116899,10.1379805,10.1515887182475,10.9950036791662,10.1857757,10.160708,10.1170373,10.6237346,10.5449929079157,8.68725904255169,10.1914604],{"iconUrl":{"data":"https://fanblairfarm.co.uk/wp-content/uploads/2020/11/chicken-icon-2.png","index":0},"iconWidth":17,"iconHeight":17},null,null,{"interactive":true,"draggable":false,"keyboard":true,"title":"","alt":"","zIndexOffset":0,"opacity":1,"riseOnHover":false,"riseOffset":250},null,null,null,null,["Nabeul","ksour essef","mellassine","bou merdes","sidi bou ali","kemicha","grombalia","Gouvernorat Siliana","Jandouba","Gouvernorat Monastir","Chela","Sijoumi","Sfax","Mahres","Stade Municipal Menzel Chaker","El Habibia","Manouba","La Marsa","تونس العاصمة","Municipalité De Beb Lakwes","Bizerte","Chela","El Aouina","mellassine","Gouvernorat Béja","Zaghouan","Bir Ali Ben Khalifa","Chela","Gouvernorat Mahdia","Hayy Ibn Khaldun","mellassine","Délégation Ksour Essef","تونس العاصمة","Rue Ibn Charaf","Intilaka","Rue Touhami Amar","Agareb","Aïn Draham","Rue Beb Jdid"],{"interactive":false,"permanent":false,"direction":"auto","opacity":1,"offset":[0,0],"textsize":"10px","textOnly":false,"className":"","sticky":true},null]},{"method":"addProviderTiles","args":["MtbMap",null,null,{"errorTileUrl":"","noWrap":false,"detectRetina":false}]},{"method":"addProviderTiles","args":["Stamen.TonerLines",null,null,{"errorTileUrl":"","noWrap":false,"opacity":0.35,"detectRetina":false}]},{"method":"addProviderTiles","args":["Stamen.TonerLabels",null,null,{"errorTileUrl":"","noWrap":false,"detectRetina":false}]}],"limits":{"lat":[34.5218412,37.2720905],"lng":[8.68725904255169,10.9987172]}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="best-map-until-now" class="section level3">
<h3>Best map until now</h3>
<pre class="r"><code>#Red chicken Tile, the best one until now

m &lt;- leaflet() %&gt;%
    addTiles(urlTemplate = &quot;https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png&quot;) %&gt;%    addMarkers(query$lon, query$lat, label = query$query,  icon = makeIcon(iconUrl =&quot;https://www.downloadclipart.net/large/chicken-png-transparent-image.png&quot;, iconWidth = 17, iconHeight= 17)) 

m %&gt;% addProviderTiles(providers$Stamen.TonerLite) #best map until now</code></pre>
<div class="leaflet html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-6534a8d720ce74e247fd" style="width:672px;height:480px;"></div>
<script type="application/json" data-for="htmlwidget-6534a8d720ce74e247fd">{"x":{"options":{"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}}},"calls":[{"method":"addTiles","args":["https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",null,null,{"minZoom":0,"maxZoom":18,"tileSize":256,"subdomains":"abc","errorTileUrl":"","tms":false,"noWrap":false,"zoomOffset":0,"zoomReverse":false,"opacity":1,"zIndex":1,"detectRetina":false}]},{"method":"addMarkers","args":[[36.4512897,35.4162835,36.79515,35.461005,35.9606191,36.5044991,36.5984231,35.9715323,36.494946,35.60547975,36.705375,36.7769937,34.7232387,34.5218412,34.9693859,36.8108651,36.7624363,36.8790882,36.8002068,36.80647135,37.2720905,36.705375,36.852609,36.79515,36.74623035,36.3319504,34.7267806,36.705375,35.3409383,36.82794,36.79515,35.35660215,36.8002068,36.7989826,36.8391913,35.8156262,34.7267589,36.77393155,36.8597626],[10.7355915,10.9987172,10.1515887182475,10.7091095,10.4821146467179,10.6766828,10.4990402,9.3577128756367,8.776369,10.7876947685139,10.2982458529412,10.1091852321074,10.3358474256388,10.4984421,10.3738282683642,9.95370955407237,9.83361907826277,10.327678,10.1857757,10.1641321520558,9.8708565,10.2982458529412,10.2659363,10.1515887182475,9.20323283692005,10.0452999718776,10.1014808309275,10.2982458529412,10.6041957116899,10.1379805,10.1515887182475,10.9950036791662,10.1857757,10.160708,10.1170373,10.6237346,10.5449929079157,8.68725904255169,10.1914604],{"iconUrl":{"data":"https://www.downloadclipart.net/large/chicken-png-transparent-image.png","index":0},"iconWidth":17,"iconHeight":17},null,null,{"interactive":true,"draggable":false,"keyboard":true,"title":"","alt":"","zIndexOffset":0,"opacity":1,"riseOnHover":false,"riseOffset":250},null,null,null,null,["Nabeul","ksour essef","mellassine","bou merdes","sidi bou ali","kemicha","grombalia","Gouvernorat Siliana","Jandouba","Gouvernorat Monastir","Chela","Sijoumi","Sfax","Mahres","Stade Municipal Menzel Chaker","El Habibia","Manouba","La Marsa","تونس العاصمة","Municipalité De Beb Lakwes","Bizerte","Chela","El Aouina","mellassine","Gouvernorat Béja","Zaghouan","Bir Ali Ben Khalifa","Chela","Gouvernorat Mahdia","Hayy Ibn Khaldun","mellassine","Délégation Ksour Essef","تونس العاصمة","Rue Ibn Charaf","Intilaka","Rue Touhami Amar","Agareb","Aïn Draham","Rue Beb Jdid"],{"interactive":false,"permanent":false,"direction":"auto","opacity":1,"offset":[0,0],"textsize":"10px","textOnly":false,"className":"","sticky":true},null]},{"method":"addProviderTiles","args":["Stamen.TonerLite",null,null,{"errorTileUrl":"","noWrap":false,"detectRetina":false}]}],"limits":{"lat":[34.5218412,37.2720905],"lng":[8.68725904255169,10.9987172]}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<div id="make-pca-for-my-data" class="section level1">
<h1>Make PCA for my data</h1>
<p>I will find this usefull to interprete my data, the thing is some of
the vectors here were manually typed</p>
<div id="loading-libraries" class="section level3">
<h3>Loading Libraries</h3>
<pre class="r"><code>library(logisticPCA)
library(ggplot2)</code></pre>
</div>
<div id="importing-the-data-and-rendering-the-graph"
class="section level3">
<h3>Importing the Data and rendering the graph</h3>
<pre class="r"><code>All_df &lt;- read.csv(file = &quot;C:/Users/User/OneDrive - Université Laval/OMICS/Paillasse/All_Data - Copy.csv&quot;) # This file contains most of the genomic data
data&lt;- subset(All_df, select = c(-gene))
Activity &lt;-c(rep(&quot;Active&quot;, 21), rep(&quot;Inactive&quot;, 18))

data &lt;- as.matrix(data) # with matrix you can have repetitive rownames
rownames(data)&lt;- Activity
logsvd_model = logisticSVD(data, k = 2)
logsvd_model</code></pre>
<pre><code>## 39 rows and 321 columns
## Rank 2 solution
## 
## 53.8% of deviance explained
## 470 iterations to converge</code></pre>
<pre class="r"><code>logpca_cv = cv.lpca(data, ks = 2, ms = 1:10)
plot(logpca_cv)</code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-27-1.png" width="672" /></p>
<pre class="r"><code>logpca_model = logisticPCA(data, k = 2, m = which.min(logpca_cv))
plot(logsvd_model, type = &quot;trace&quot;)</code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-27-2.png" width="672" /></p>
<pre class="r"><code>Activity&lt;- rownames(data)
plot(logpca_model, type = &quot;scores&quot;) + geom_point(aes(colour = Activity)) + 
  ggtitle(&quot;Logistic PCA&quot;) + scale_colour_manual(values = c(&quot;blue&quot;, &quot;red&quot;)) </code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-27-3.png" width="672" /></p>
<pre class="r"><code>plot(logsvd_model, type = &quot;scores&quot;) + geom_point(aes(colour = Activity)) +   ggtitle(&quot;Exponential Family PCA&quot;) + scale_colour_manual(values = c(&quot;blue&quot;, &quot;red&quot;))</code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-27-4.png" width="672" /></p>
</div>
<div
id="pca-coloration-depending-on-the-criteria-of-bacteriocin-production"
class="section level3">
<h3>PCA coloration depending on the criteria of bacteriocin
Production</h3>
<p>Here we will use the same model, the coordinates of the population
will stay in the same place no need to redo the calculation. However we
will colour the strains depending on the criterium of bacteriocin
production</p>
<pre class="r"><code>Mcc &lt;- c(&quot;J25&quot;, &quot;J25+C7+colV&quot;, &quot;J25+colV&quot;, &quot;J25&quot;, &quot;J25+colV&quot;, &quot;inactive&quot;, &quot;inactive&quot;, &quot;inactive&quot;, &quot;colV&quot;, &quot;inactive&quot;, &quot;C7+colV&quot;, &quot;colV&quot;, &quot;inactive&quot;, &quot;colV&quot;, &quot;inactive&quot;, &quot;colV&quot;, &quot;colV&quot;, &quot;inactive&quot;, &quot;colV&quot;, &quot;inactive&quot;, &quot;inactive&quot;, &quot;colV&quot;, rep(&quot;inactive&quot;, 5), rep(&quot;colV&quot;, 2), rep(&quot;inactive&quot;, 7), &quot;colV&quot;, &quot;inactive&quot;, &quot;inactive&quot;)
Mcc[Mcc==&quot;inactive&quot;] &lt;- &quot;Non-producer&quot;
Mcc[Mcc!=&quot;Non-producer&quot;] &lt;- &quot;Bacteriocin producer&quot;


plot(logsvd_model, type = &quot;scores&quot;) + geom_point(aes(colour = Mcc)) + 
  ggtitle(&quot;Exponential Family PCA&quot;) + scale_colour_manual(values = c(&quot;blue&quot;, &quot;orange&quot;))</code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-28-1.png" width="672" /></p>
<pre class="r"><code>p1 &lt;- plot(logpca_model, type = &quot;scores&quot;) + geom_point(aes(colour = Mcc)) +   ggtitle(&quot;Logistic PCA&quot;) + scale_colour_manual(values = c(&quot;blue&quot;, &quot;orange&quot;)) 

p1 #visualize the plot</code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-28-2.png" width="672" /></p>
</div>
<div id="pca-depending-on-the-degree-of-virulence"
class="section level3">
<h3>PCA depending on the degree of Virulence</h3>
<p>Same thing but different colours depending on the pathotype.</p>
<pre class="r"><code>#Virulence&lt;- dput(as.character(clipr::read_clip())) # very useful to copy paste a vector 

Virulence&lt;- c(&quot;APEC potential&quot;, &quot;APEC&quot;, &quot;APEC potential&quot;, &quot;APEC potential&quot;, &quot;APEC potential&quot;, &quot;AFEC&quot;, &quot;AFEC&quot;, &quot;APEC&quot;, &quot;APEC&quot;, &quot;APEC potential&quot;, &quot;APEC&quot;, &quot;APEC potential&quot;, &quot;APEC&quot;, &quot;APEC&quot;, &quot;APEC potential&quot;, &quot;APEC&quot;, &quot;APEC&quot;, &quot;APEC potential&quot;, &quot;APEC&quot;, &quot;APEC potential&quot;, &quot;APEC potential&quot;, &quot;APEC&quot;, &quot;APEC potential&quot;, &quot;AFEC&quot;, &quot;AFEC&quot;, &quot;APEC&quot;, &quot;AFEC&quot;, &quot;APEC&quot;, &quot;APEC&quot;, &quot;AFEC&quot;, &quot;AFEC&quot;, &quot;AFEC&quot;, &quot;AFEC&quot;, &quot;AFEC&quot;, &quot;AFEC&quot;, &quot;AFEC&quot;,
&quot;APEC&quot;, &quot;AFEC&quot;, &quot;APEC potential&quot;)

plot(logsvd_model, type = &quot;scores&quot;) + geom_point(aes(colour = Virulence)) + 
  ggtitle(&quot;Exponential Family PCA&quot;) + scale_colour_manual(values = c(&quot;green&quot;, &quot;red&quot;, &quot;purple&quot;))</code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-29-1.png" width="672" /></p>
<pre class="r"><code>p2 &lt;- plot(logpca_model, type = &quot;scores&quot;) + geom_point(aes(colour = Virulence)) + 
  ggtitle(&quot;Logistic PCA&quot;) + scale_colour_manual(values = c(&quot;green&quot;, &quot;red&quot;, &quot;purple&quot;)) 

p2 #visualize the plot </code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-29-2.png" width="672" /></p>
</div>
<div id="pca-depending-on-phylogroup" class="section level3">
<h3>PCA depending on Phylogroup</h3>
<pre class="r"><code>#Phylogroup &lt;- dput(as.character(clipr::read_clip()))
Phylogroup &lt;- c(&quot;A&quot;, &quot;A&quot;, &quot;F&quot;, &quot;A&quot;, &quot;B1&quot;, &quot;A&quot;, &quot;A&quot;, &quot;B1&quot;, &quot;B1&quot;, &quot;A&quot;, &quot;B1&quot;, &quot;A&quot;, &quot;B1&quot;, &quot;B1&quot;, &quot;A&quot;, &quot;F&quot;, &quot;B1&quot;, &quot;B2&quot;, &quot;D&quot;, &quot;B1&quot;, &quot;A&quot;, &quot;B1&quot;, &quot;B1&quot;, &quot;A&quot;, &quot;A&quot;, &quot;B1&quot;, &quot;A&quot;, &quot;B1&quot;, &quot;B1&quot;, &quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;D&quot;)

plot(logsvd_model, type = &quot;scores&quot;) + geom_point(aes(colour = Phylogroup)) +   ggtitle(&quot;Exponential Family PCA&quot;) + scale_colour_manual(values = c(&quot;green&quot;, &quot;red&quot;, &quot;purple&quot;, &quot;yellow&quot;, &quot;grey&quot;))</code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-30-1.png" width="672" /></p>
<pre class="r"><code>p3 &lt;- plot(logpca_model, type = &quot;scores&quot;) + geom_point(aes(colour = Phylogroup)) +   ggtitle(&quot;Logistic PCA&quot;) + scale_colour_manual(values = c(&quot;green&quot;, &quot;red&quot;, &quot;purple&quot;, &quot;yellow&quot;, &quot;grey&quot;)) 

p3 # visualize the plot</code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-30-2.png" width="672" /></p>
<p>#Merging the plots</p>
</div>
<div id="loadin-the-patchwork-package" class="section level3">
<h3>Loadin the Patchwork package</h3>
<pre class="r"><code>library(patchwork)
p1+p2+p3</code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-31-1.png" width="672" /></p>
</div>
</div>
<div id="usefull-stuff" class="section level1">
<h1>Usefull stuff</h1>
<div id="merging-two-tables-by-their-id" class="section level3">
<h3>merging two tables by their ID</h3>
<p>Let’s suppose that we have two tables that contain the same unique
IDs, this function is usefull to merge the two tables depending on these
IDs even if the rows are not ordered. Here we merge two tables data1 and
data2 by their ID names</p>
<pre class="r"><code>merge(x= data1, y=data2, by= &quot;ID&quot;, all=TRUE)</code></pre>
</div>
<div id="replace-all-nas-with-0-in-a-dataframe" class="section level3">
<h3>Replace all NA’s with 0 in a dataframe</h3>
<p>Here we replace all the NAs in a data by zero, easy stuff</p>
<pre class="r"><code>data[is.na(data)] &lt;- 0</code></pre>
</div>
<div id="calculate-correlation-for-binary-table" class="section level3">
<h3>Calculate correlation for binary table</h3>
<p>Here i want to calculate the pearson’s correlation coefficient
between my different variables. By checking the bibliography, the cor()
function should do the work. If there is a variable that takes the value
1 or 0 across all the population, we will see this Warning message : In
cor(data) : the standard deviation is zero We can make a heatmap Based
on that using the pheatmap package</p>
<pre class="r"><code>cor_data &lt;- cor(data)
library(pheatmap)
cor_data[is.na(cor_data)] &lt;- 0
pheatmap::pheatmap(cor_data)</code></pre>
</div>
</div>
<div id="plasmids" class="section level1">
<h1>Plasmids</h1>
<p>So I have a Table containing some presence/absence Data of some
plasmids across my genomes.</p>
<div id="import-data-1" class="section level3">
<h3>Import Data</h3>
<p>The First time that I worked with my data, i just used clipr package
to get it from the clipboard, then i saved it to a file that i will read
the second time.</p>
<pre class="r"><code>#tb2 &lt;-clipr::read_clip_tbl(header=TRUE)

#write.csv(tb2 , row.names = FALSE, file = &quot;C:/Users/User/OneDrive - Université Laval/OMICS/Paillasse/Plasmids_binary.csv&quot;)

tb2 &lt;- read.csv(file = &quot;C:/Users/User/OneDrive - Université Laval/OMICS/Paillasse/Plasmids_binary.csv&quot;)</code></pre>
</div>
<div id="replace-nas" class="section level3">
<h3>Replace NAs</h3>
<p>It’s as easy as this</p>
<pre class="r"><code>tb2[is.na(tb2)== TRUE]&lt;-0</code></pre>
</div>
</div>
<div id="phylogenetic-tree" class="section level1">
<h1>Phylogenetic Tree</h1>
<div id="load-ape-package" class="section level3">
<h3>Load ape Package</h3>
<pre class="r"><code>library(ape)</code></pre>
</div>
<div id="import-newick-file" class="section level3">
<h3>Import newick file</h3>
<pre class="r"><code>mytree&lt;-read.tree(&quot;C:/Users/User/OneDrive - Université Laval/OMICS/Paillasse/tree.newick.txt&quot;)

mytree$tip.label # to check the edge number of the APEC strain</code></pre>
<pre><code>##  [1] &quot;Sample36&quot;                                                 
##  [2] &quot;Sample34&quot;                                                 
##  [3] &quot;Sample30&quot;                                                 
##  [4] &quot;Sample35&quot;                                                 
##  [5] &quot;Sample10&quot;                                                 
##  [6] &quot;Sample7&quot;                                                  
##  [7] &quot;Sample6&quot;                                                  
##  [8] &quot;Sample2&quot;                                                  
##  [9] &quot;Sample4&quot;                                                  
## [10] &quot;Sample1&quot;                                                  
## [11] &quot;Sample31&quot;                                                 
## [12] &quot;Sample27&quot;                                                 
## [13] &quot;Sample38&quot;                                                 
## [14] &quot;Sample21&quot;                                                 
## [15] &quot;Sample12&quot;                                                 
## [16] &quot;Sample37&quot;                                                 
## [17] &quot;Sample25&quot;                                                 
## [18] &quot;Sample33&quot;                                                 
## [19] &quot;Sample32&quot;                                                 
## [20] &quot;Sample15&quot;                                                 
## [21] &quot;Sample24&quot;                                                 
## [22] &quot;Sample41&quot;                                                 
## [23] &quot;NC_000913_3_Escherichia_coli_str_K_12_substr_MG1655_fixed&quot;
## [24] &quot;Sample14&quot;                                                 
## [25] &quot;Sample13&quot;                                                 
## [26] &quot;Sample17&quot;                                                 
## [27] &quot;Sample26&quot;                                                 
## [28] &quot;Sample23&quot;                                                 
## [29] &quot;Sample11&quot;                                                 
## [30] &quot;Sample20&quot;                                                 
## [31] &quot;Sample22&quot;                                                 
## [32] &quot;Sample40&quot;                                                 
## [33] &quot;NC_020163_1_Escherichia_coli_APEC_O78_fixed&quot;              
## [34] &quot;NC_018658_1_Escherichia_coli_O104_H4_str_2011C_3493_fixed&quot;
## [35] &quot;Sample9&quot;                                                  
## [36] &quot;Sample29&quot;                                                 
## [37] &quot;Sample5&quot;                                                  
## [38] &quot;Sample28&quot;                                                 
## [39] &quot;Sample8&quot;                                                  
## [40] &quot;Sample18&quot;                                                 
## [41] &quot;NC_017634_1_Escherichia_coli_O83_H1_str_NRG_857C_fixed&quot;   
## [42] &quot;Sample3&quot;                                                  
## [43] &quot;Sample16&quot;                                                 
## [44] &quot;NC_011750_1_Escherichia_coli_IAI39_fixed&quot;                 
## [45] &quot;Sample19&quot;                                                 
## [46] &quot;CU928163_2_Escherichia_coli_UMN026_fixed&quot;                 
## [47] &quot;Sample39&quot;                                                 
## [48] &quot;NC_002695_2_Escherichia_coli_O157_H7_str_Sakai_fixed&quot;</code></pre>
<pre class="r"><code>rerootTree&lt;- root(mytree, mytree$tip.label[33])

rerootTree$tip.label </code></pre>
<pre><code>##  [1] &quot;Sample36&quot;                                                 
##  [2] &quot;Sample34&quot;                                                 
##  [3] &quot;Sample30&quot;                                                 
##  [4] &quot;Sample35&quot;                                                 
##  [5] &quot;Sample10&quot;                                                 
##  [6] &quot;Sample7&quot;                                                  
##  [7] &quot;Sample6&quot;                                                  
##  [8] &quot;Sample2&quot;                                                  
##  [9] &quot;Sample4&quot;                                                  
## [10] &quot;Sample1&quot;                                                  
## [11] &quot;Sample31&quot;                                                 
## [12] &quot;Sample27&quot;                                                 
## [13] &quot;Sample38&quot;                                                 
## [14] &quot;Sample21&quot;                                                 
## [15] &quot;Sample12&quot;                                                 
## [16] &quot;Sample37&quot;                                                 
## [17] &quot;Sample25&quot;                                                 
## [18] &quot;Sample33&quot;                                                 
## [19] &quot;Sample32&quot;                                                 
## [20] &quot;Sample15&quot;                                                 
## [21] &quot;Sample24&quot;                                                 
## [22] &quot;Sample41&quot;                                                 
## [23] &quot;NC_000913_3_Escherichia_coli_str_K_12_substr_MG1655_fixed&quot;
## [24] &quot;Sample14&quot;                                                 
## [25] &quot;Sample13&quot;                                                 
## [26] &quot;Sample17&quot;                                                 
## [27] &quot;Sample26&quot;                                                 
## [28] &quot;Sample23&quot;                                                 
## [29] &quot;Sample11&quot;                                                 
## [30] &quot;Sample20&quot;                                                 
## [31] &quot;Sample22&quot;                                                 
## [32] &quot;Sample40&quot;                                                 
## [33] &quot;NC_020163_1_Escherichia_coli_APEC_O78_fixed&quot;              
## [34] &quot;NC_018658_1_Escherichia_coli_O104_H4_str_2011C_3493_fixed&quot;
## [35] &quot;Sample9&quot;                                                  
## [36] &quot;Sample29&quot;                                                 
## [37] &quot;Sample5&quot;                                                  
## [38] &quot;Sample28&quot;                                                 
## [39] &quot;Sample8&quot;                                                  
## [40] &quot;Sample18&quot;                                                 
## [41] &quot;NC_017634_1_Escherichia_coli_O83_H1_str_NRG_857C_fixed&quot;   
## [42] &quot;Sample3&quot;                                                  
## [43] &quot;Sample16&quot;                                                 
## [44] &quot;NC_011750_1_Escherichia_coli_IAI39_fixed&quot;                 
## [45] &quot;Sample19&quot;                                                 
## [46] &quot;CU928163_2_Escherichia_coli_UMN026_fixed&quot;                 
## [47] &quot;Sample39&quot;                                                 
## [48] &quot;NC_002695_2_Escherichia_coli_O157_H7_str_Sakai_fixed&quot;</code></pre>
<pre class="r"><code>plot(rerootTree)</code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-38-1.png" width="672" /></p>
<pre class="r"><code>edgecol&lt;- vector()
edgecol[1:41] = &quot;black&quot;

tipcol&lt;- vector()
tipcol[1:41] = &quot;black&quot;

arbre &lt;- plot.phylo(rerootTree, &quot;tidy&quot;,  main= &quot;phylogenetic analysis based on Average nucleotide identity&quot;, edge.color =  edgecol, cex = 0.6 , tip.color = tipcol, use.edge.length = FALSE, x.lim = 90) </code></pre>
<p><img src="Journal_files/figure-html/unnamed-chunk-38-2.png" width="672" /></p>
</div>
<div id="adding-labels-optionnal" class="section level3">
<h3>Adding labels (optionnal)</h3>
<pre class="r"><code>#edgelabels(font=0.5, cex=0.5) 
#tiplabels(font=0.05, cex=0.5)</code></pre>
</div>
<div id="chi-square-test-on-binary-data" class="section level3">
<h3>Chi Square test on binary data</h3>
<p>credit goes to <a
href="https://stackoverflow.com/questions/71666349/multiple-chi-square-tests-in-r"
class="uri">https://stackoverflow.com/questions/71666349/multiple-chi-square-tests-in-r</a></p>
<pre class="r"><code>data &lt;- read.csv(file = &quot;C:/Users/User/OneDrive - Université Laval/OMICS/Paillasse/All_Data - Copy.csv&quot;, stringsAsFactors = TRUE) # This file contains most of the genomic data 

# Identify columns with only one level as a value
cols_to_remove &lt;- sapply(data, function(x) length(unique(x)) == 1)

#Remove columns with only one level as a value
data &lt;- data[, !cols_to_remove]

# Split the data into response (sample) and predictor (parameter) variables
#response &lt;- data[,1]
#predictors &lt;- data[,2:ncol(data)]

#response &lt;- as.factor(response) #necesary for chisq.test to work

# Use the chi-squared test to determine the association between each parameter
chi_sq_test &lt;- data %&gt;% 
    select(-gene) %&gt;% 
    colnames() %&gt;% 
    combn(2) %&gt;% 
    t() %&gt;% 
    as_tibble() %&gt;% 
    rowwise %&gt;% 
    mutate(chisq_test = list(
        table(data[[V1]], data[[V2]]) %&gt;% chisq.test()
    ),
    chisq_pval = chisq_test$p.value
    )

# Sort the chi-squared test statistics in descending order to determine the most discriminatory parameters
sorted_chi_sq_test &lt;- chi_sq_test[order(chi_sq_test$chisq_pval, decreasing = FALSE), ]

# Identify the top 10 most discriminatory parameters
top_10_params &lt;- head(sorted_chi_sq_test, 10)

# Print the top 10 most discriminatory parameters
print(top_10_params)</code></pre>
<pre><code>## # A tibble: 10 × 4
## # Rowwise: 
##    V1        V2          chisq_test    chisq_pval
##    &lt;chr&gt;     &lt;chr&gt;       &lt;list&gt;             &lt;dbl&gt;
##  1 APH.6..Id APH.3....Ib &lt;htest&gt;    0.00000000316
##  2 gspC      gspE        &lt;htest&gt;    0.00000000334
##  3 gspC      gspF        &lt;htest&gt;    0.00000000334
##  4 gspE      gspF        &lt;htest&gt;    0.00000000334
##  5 gspH      gspI        &lt;htest&gt;    0.00000000349
##  6 gspH      gspJ        &lt;htest&gt;    0.00000000349
##  7 gspH      gspK        &lt;htest&gt;    0.00000000349
##  8 gspI      gspJ        &lt;htest&gt;    0.00000000349
##  9 gspI      gspK        &lt;htest&gt;    0.00000000349
## 10 gspJ      gspK        &lt;htest&gt;    0.00000000349</code></pre>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
